### DISCUSSION TITLE
How to do performance debugging?

### DISCUSSION THREAD

**Discussion by mavam** [ArthurSonzogni/FTXUI#698] [Q&A] [answered]
I'm trying to understand why my application runs so incredibly slow. I did some stack sampling but don't get quite far:

<img width="1614" alt="image" src="https://github.com/ArthurSonzogni/FTXUI/assets/53797/68087512-344d-4ee5-9565-fa376f02f61a">

(This is [speedscope](https://www.speedscope.app/) and [here's the sample](https://github.com/ArthurSonzogni/FTXUI/files/12063831/tmp-22211.folded.zip).)

I'm looking at potentially really large tables, like this one:

<img width="1393" alt="image" src="https://github.com/ArthurSonzogni/FTXUI/assets/53797/4dddc60d-994d-4f17-8fc3-481c6f8d06fc">

I think the functional reactive UI makes it hard to narrow this down to a root cause. I have several hunches, of course, but I cannot validate them one-by-one efficiently. For example, the columns (with potentially 10k elements each) are being added like this:

```cpp
        Elements body;
        body.reserve(slice.rows());
        for (size_t row = 0; row < slice.rows(); ++row) {
          auto cell = StaticCell(slice.at(row, col), state_->theme);
          body.push_back(std::move(cell));
        }
        body_->Add(lift(vbox(std::move(body))));
```

I can only guess that this is somewhere visible on the flame graph.

How do you do performance debugging?

**Comment by ArthurSonzogni** [comment, marked as answer]
It looks like it spent time trying to determine the size of string.
Could you check if https://github.com/ArthurSonzogni/FTXUI/pull/691 fixed the issue?

Then I would try to limit the amount of lines rendered inside the table. Most of it won't be shown, so there are no need to do the layout for 10k elements.
I see you took a `slice`, so that's probably what you are already doing.

**Reply by mavam** [reply]
> Could you check if https://github.com/ArthurSonzogni/FTXUI/pull/691 fixed the issue?

It did!

> I see you took a `slice`, so that's probably what you are already doing.

Right, table slices are still potentially large, and the number of components grows quadratically with the number of columns, so > 10k components isn't unusual.

**Reply by mavam** [reply]
Any chance you could cut a minor or patch release that fixes this performance issue? It would help us a lot in our packaging.

**Reply by ArthurSonzogni** [reply]
This sounds like a good idea.
I cherry-picked a dozen of patch into the `support-v4` branch to prepare the `v4.2`. Could you please double check this looking good to you?

**Reply by mavam** [reply]
I'm getting a few errors on that branch, e.g.,:

```
error: use of undeclared identifier 'Left'; did you mean 'ftxui::Mouse::Left'?
      case Left:
           ^~~~
           ftxui::Mouse::Left
.../aux/ftxui/include/ftxui/component/mouse.hpp:10:5: note: 'ftxui::Mouse::Left' declared here
    Left = 0,
               ^
error: use of undeclared identifier 'Up'; did you mean 'ftxui::MenuOption::Up'?
      case Up:
           ^~
           ftxui::MenuOption::Up
```

Maybe a few more commits are missing?

**Reply by ArthurSonzogni** [reply]
Are you sure? I can't reproduce.

Could you try again? What file is failing to compile? I don't have any `case Left:` occurrence in my library.
I don't see any API breakage when comparing v4.2 vs v4.1.1

**Reply by mavam** [reply]
The `Left` error comes up from a small helper function where I have `using enum Direction`. The full qualified token would be `Direction::Left`. The type `Direction` was introduced with #583, which is not part of the support branch.

So while this might not be a breaking change but an enhancement, I had to refactor my application to work with main.

Could it make sense to cut a 5.0-rc1 and label that as pre-release explicitly?

**Reply by ArthurSonzogni** [reply]
I see! So you don't need a minor release from v4.1.1, but a major release.

I would like to merge the few pending PR we have.
Let's say in a week for a v5.0 ?

**Reply by ArthurSonzogni** [reply]
I see! So you don't need a minor release from v4.1.1, but a major release.

I would like to merge the few pending PR we have. One improves screen drawing benchmark by 20%.

Let's say in a week for a v5.0 ?

**Reply by mavam** [reply]
That works great timing-wise, looking forward to it!
